File: remote/rhoconnect-server/models/ruby/product.rb
require 'json'
require 'rest_client'

class Product < Rhoconnect::Model::Base
  def initialize(source)
    # Backend rails server is running locally:
    #@base = 'http://localhost:3000/products'

    # Backend rails server is running omn Heroku:
    @base = 'http://rhostore.herokuapp.com/products'
    super(source)
  end

  def login
    puts "Product#login ..."
  end

  def query(params=nil)
    rest_result = RestClient.get("#{@base}.json").body
    raise Rhoconnect::Model::Exception.new("Error connecting!") if rest_result.code != 200
    parsed = JSON.parse(rest_result)

    puts "Product#query: "
    puts parsed
    # =>
    # {"product"=>{"brand"=>"Samsung", "created_at"=>"2014-03-13T00:23:09Z", "id"=>1, "name"=>"Galaxy S5", "price"=>"$199.99", "quantity"=>"100", "sku"=>"ABC", "updated_at"=>"2014-03-13T00:34:07Z"}}
    # {"product"=>{"brand"=>"Apple", "created_at"=>"2014-03-13T00:27:28Z", "id"=>2, "name"=>"iPhone 5S ", "price"=>"$299.99", "quantity"=>"11", "sku"=>"XYZ", "updated_at"=>"2014-03-13T00:27:28Z"}}
    # {"product"=>{"brand"=>"Google", "created_at"=>"2014-03-13T00:35:33Z", "id"=>3, "name"=>"Nexus 5", "price"=>"$99.99", "quantity"=>"5", "sku"=>"POIUYT", "updated_at"=>"2014-03-13T00:35:33Z"}}


    @result = {}
    parsed.each do |item|
      @result[item["product"]["id"].to_s] = item["product"]
    end if parsed
    puts "@result: "
    puts @result
    # =>
    # {"1"=>{"brand"=>"Samsung", "created_at"=>"2014-03-13T00:23:09Z", "id"=>1, "name"=>"Galaxy S5", "price"=>"$199.99", "quantity"=>"100", "sku"=>"ABC", "updated_at"=>"2014-03-13T00:34:07Z"},
    #  "2"=>{"brand"=>"Apple", "created_at"=>"2014-03-13T00:27:28Z", "id"=>2, "name"=>"iPhone 5S ", "price"=>"$299.99", "quantity"=>"11", "sku"=>"XYZ", "updated_at"=>"2014-03-13T00:27:28Z"},
    #  "3"=>{"brand"=>"Google", "created_at"=>"2014-03-13T00:35:33Z", "id"=>3, "name"=>"Nexus 5", "price"=>"$99.99", "quantity"=>"5", "sku"=>"POIUYT", "updated_at"=>"2014-03-13T00:35:33Z"}}
  end

  def create(create_hash)
    puts "Product#create: create_hash: #{create_hash}"
    # =>
    # Product.create: create_hash: {"brand"=>"Google", "name"=>"Nexus 5", "price"=>"$99.99", "quantity"=>"5", "sku"=>"POIUYT"}

    res = RestClient.post(@base, :product => create_hash)
    puts res
    # =>
    # <?xml version="1.0" encoding="UTF-8"?>
    # <product>
    #   <brand>Google</brand>
    #   <created-at type="datetime">2014-03-13T00:35:33Z</created-at>
    #   <id type="integer">3</id>
    #   <name>Nexus 5</name>
    #   <price>$99.99</price>
    #   <quantity>5</quantity>
    #   <sku>POIUYT</sku>
    #   <updated-at type="datetime">2014-03-13T00:35:33Z</updated-at>
    # </product>
    puts "#{res.headers[:location]}.json"
    # =>
    # http://localhost:3000/products/3.json

    # After create we are redirected to the new record.
    # We need to get the id of that record and return
    # it as part of create so rhoconnect can establish a link
    # from its temporary object on the client to this newly
    # created object on the server
    body = RestClient.get("#{res.headers[:location]}.json").body
    puts "body: #{body}"
    # =>
    # body: {"product":{"brand":"Google","created_at":"2014-03-13T00:35:33Z","id":3,"name":"Nexus 5",
    #  "price":"$99.99","quantity":"5","sku":"POIUYT","updated_at":"2014-03-13T00:35:33Z"}}

    # JSON.parse(
    #     RestClient.get("#{res.headers[:location]}.json").body
    #   )["product"]["id"]
    JSON.parse(body)["product"]["id"]
  end

  def update(update_hash)
    puts "Product#update: #{update_hash}"
    # =>
    # Product.update: {"quantity"=>"100", "id"=>"1"}
    obj_id = update_hash['id']
    update_hash.delete('id')
    RestClient.put("#{@base}/#{obj_id}",:product => update_hash)
  end

  def delete(delete_hash)
    puts "Product#delete: #{delete_hash}"
    # =>
    # {"brand"=>"...", "created_at"=>"2014-03-13T21:41:55Z", "name"=>"...", "price"=>"...", "quantity"=>"...", "sku"=>"...", "updated_at"=>"2014-03-13T21:41:55Z", "id"=>"4"}

    RestClient.delete("#{@base}/#{delete_hash['id']}")
  end

  def logoff
    puts "Product#logoff."
  end

  def store_blob(object,field_name,blob)
    # TODO: Handle post requests for blobs here.
    # make sure you store the blob object somewhere permanently
    raise "Please provide some code to handle blobs if you are using them."
  end
end

File: local-without-plugin/js_rhoapp/models/js/product.js
var rc = require('rhoconnect_helpers');

var http = require('http');
var host = 'rhostore.herokuapp.com';
// var host = '192.168.1.12:3000';

var Product = function(){

  this.login = function(resp){
    console.log("**************** Login");
    resp.send(true);
  };

  this.query = function(resp){
    var result = {};
    var str = '';

    // Print message with timestamp
    console.log("**** query: " + new Date());
    http.request('http://' + host + '/products.json', function(res){
      res.on('data', function(chunk){
        // Print message with timestamp
        console.log("**** on data: " + new Date());
        //
        str += chunk;
      });
      res.on('end', function(){
        var data = JSON.parse(str);
        for(var i in data){
          var item = data[i];
          result[item.product.id.toString()] = item.product;
        }
        // ************************
        console.log("**** on end: "  + new Date());
        console.log(result);
        // ************************
        resp.send(result);
      });
      res.on('error', function(){
        console.log("**** on error: "  + new Date());
        resp.send(false);
      });
    }).end();
  };

  this.create = function(resp){
    var postData = JSON.stringify({ 'product': resp.params.create_object });
    var str = '';
    var options = {
      host: host,
      path: '/products.json',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': postData.length
      }
    };
    var req = http.request(options, function(res){
      res.on('data', function(chunk){
        str += chunk;
      });
      res.on('end', function(){
        var data = JSON.parse(str);
        resp.send(data.product.id.toString());
      });
    });
    req.write(postData);
    req.end();
  };

  this.update = function(resp){
    var objId = resp.params.update_object.id;
    var putData = JSON.stringify({ "product": resp.params.update_object });
    // Remove the id from the hash, we don't need it.
    delete putData.id;
    var options = {
      host: host,
      path: '/products/' + objId + '.json',
      method: 'PUT',
      headers: {
      'Content-Type': 'application/json',
      'Content-Length': putData.length
      }
    };
    var req = http.request(options, function(res){
        res.on('data', function(){});
        res.on('end', function(){
        resp.send(true);
      });
      res.on('error', function(){
        resp.send(false);
      });
    });
    req.write(putData);
    req.end();
  };

  this.del = function(resp){
    var objId = resp.params.delete_object.id;
    var options = {
      host: host,
      path: '/products/' + objId + '.json',
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' }
    };
    var req = http.request(options, function(res){
      res.on('data', function(){});
      res.on('end', function(){
        resp.send(true);
      });
      res.on('error', function(){
        resp.send(false);
      });
    });
    req.end();
  };

  this.logoff = function(resp){
    console.log("**************** Logoff");
    resp.send(true);
  };
};

module.exports = new Product();
